controllers/project.lua

```diff
@@ -298,15 +279,12 @@ ProjectController = {
             -- Parameters:  page, pagesize
             exptime = 60, -- cache expires after 60 seconds
             function (self)
-                local project =
-                    Projects:find(self.params.username, self.params.projectname)
+                local project = assert_project(self, self.params.username, self.params.projectname)
 
-                if not project then yield_error(err.nonexistent_project) end
-                if not project.ispublic then
-                    assert_users_match(self, err.nonexistent_project)
-                end
+                assert_error(can_view_project(self, project), err.nonexistent_project)
 
                 -- This logic is extremely convoluted. It needs to be rethought.
+                -- collections = Project:get_collections() then filter with can_view_collection()
```

Comment:
I think it's best to move some of this logic to the server.. I don't think the server-side filtering will make a difference in performance.

I did realize we could remove the `JOIN` on users, though.